### PR TITLE
Addressing PR comments

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: PR Comments (sub-batch)

**Motivation and Context**
- https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397642684
- https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397646584
- https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397643765
- https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397019190
- https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397023575
- https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397028820
- https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397648046

### CODE REVIEW

File: onnxruntime/core/framework/session_state.cc [hunk 1]
```diff
@@ -228,7 +227,7 @@ Status SessionState::GeneratePatternGroupCache(const std::vector<std::reference_
     auto* node = graph_viewer_->GetNode(node_plan.node_index);
     int output_start = node_index + static_cast<int>(node->InputDefs().size()) + static_cast<int>(node->ImplicitInputDefs().size());
     //allocate output
-    for (int i = 0; i < static_cast<int>(node->OutputDefs().size()); ++i) {
+    for (int i = 0, end = static_cast<int>(node->OutputDefs().size()); i < end; ++i) {
```
Comment (edgchen1) [inline onnxruntime/core/framework/session_state.cc]: can we just use size_t and avoid the cast to int here?

Comment (ytaous) [reply onnxruntime/core/framework/session_state.cc]: I think so, let me change that too

File: onnxruntime/core/framework/session_state.cc [hunk 2]
```diff
@@ -252,7 +251,9 @@ Status SessionState::GeneratePatternGroupCache(const std::vector<std::reference_
             }
             len *= it->second;
           } else {
-            len *= dim.dim_value();
+            if (dim.has_dim_value()) {
```
Comment (edgchen1) [inline onnxruntime/core/framework/session_state.cc]: should we error out if neither dim_value nor dim_param is set? that means it's unknown: https://github.com/onnx/onnx/blob/master/docs/IR.md#static-tensor-shapes

Comment (ytaous) [reply onnxruntime/core/framework/session_state.cc]: sure

**Pull request merged**